### ISSUE TITLE
PMAPs interface

### ISSUE THREAD

**Issue by jacg** [next-exp/IC#177]
The PMAPs interface we currently have is an accidental by-product of the implementation choices we happen to be going with right now: no effort has been made to design a decent interface for PMAPs.

The presence or absence of zeros (whether it be in the HDF5 file or in the datastructure used in the transient representation) in the S2Sis is a matter of implementation: the client doesn't care how and where the information is stored, as long as it is accessible conveniently and quickly.

Similarly, whether times or sample numbers are stored in the energy plane components of the pmaps and/or the tracking plane components, is purely a detail of implementation: the client just needs to access information as quickly and conveniently as possible.

Our choice to store times and energies in the S12s, but only energies in the S2Sis, is an implementation detail which we currently throw into the client's face.This choice of implementation *leaks out through the interface*: Compare and contrast how to access energies and their corresponding times in the different planes, when you are holding on to some peak `p`:

+ S12s:
    - energy: `p.E[n]` OK.
    - time  `p.t[n]` Nice symmetry.
+ S2Sis:
    - energy: `p[n]` Eh? Where did the `.E` go?
    - time:  erm ... `a_related_but_different_p[event_no].t[n]`? WTF!

This is an *abominable*, inconsistent interface with a horrible abstraction leak. It is like it is because the implementors (that's us!) haven't bothered to think about the *client* (that's also us!) interface. It needs to be killed with fire!

What needs to be done:

+ [ ] Collect use cases, showing the sorts of things we expect clients to do with PMAPs.
+ [ ] Design interface, one which allows the client to use PMAPs cleanly, efficiently and consistently.
+ [ ] Hide implementation behind interface.

If y'all provide me with use cases, I can tackle the rest in April. Of course, this is best done *con peras*, but the collection of use cases can start right now. The collection of use cases is also the part where I can probably contribute least.

There is lots more to be said on this topic, but this should be enough for a start.

**Comment by jjgomezcadenas** [comment]
>This is an abominable, inconsistent interface with a horrible abstraction leak.

So say we all! As a physicist I am shocked to find out that the what I thought as the prettiest of all the interfaces is plain crap. As a wizard's apprentice I am happy to eat peras with you as soon as possible and have some fun designing and implementing a shiny JCK interface.

We all should take notice of that and starting right now think the interfaces before we start implementing. 

Thanks Master

**Comment by jacg** [comment]
I hope that y'all have collected lots of juicy use cases for me.

I expect my inbox to be flooded with them any minute now!

**Comment by jjgomezcadenas** [comment]
More peras, next week!

**Comment by jjgomezcadenas** [comment]
Solved by new interface

**Comment by jacg** [comment]
As it was unilaterally closed without the planned and promised gathering and analysis of use cases, without the promised peras, and given the fact that there is wide dissatisfaction with the interface as it stands, this issue is being re-opened.

After extensive discussion with @abotas, @Aretno, Carmen (whose GitHub
handle I can't find at the moment) and @gonzaponte, we have arrived at
the following proposal for a PMAPs interface.

The current `S1`, `S2`, `S1pmt`, `S2pmt`, `S2si`, `S12` and `Peak` to
be replaced with 3 classes, with at least the following features in
their interfaces:


    class Signal:
    
        peaks                 -> Iterator[Peak]
        n_peaks               :  int
        get_peak(peak_number) -> Peak
    
    class Peak:
    
        sipms : SensorResponses
        pmts  : SensorResponses
        times : Sequence
    
    class SensorResponses:
    
        all_waveforms       : 2D array
        waveform(sensor_id) : 1D array
        time_slice(index)   : 1D array 
        ids                 : Sequence[int]
        sum_over_times      : 1D array
        sum_over_sensors    : 1D array

    

More explicitly,

1. a `Signal` contains a number of `Peak`s and

   + can tell us how many peaks it contains
   + can give us any single peak, specified by its number
   + can give us an iterator over all its peaks

2. a `Peak` contains

   + sensor data for the SiPMs
   + sensor data for the PMTs
   + a sequence of times at which the amplitudes in the sensors were
     observed

3. a collection of sensor data for the peak (grouped separately for the PMTs and the SiPMs) can be queried in various ways

   + 2D array of all sensor amplitudes for all times slices for all
     sensors. In the case of the SiPMs, only those sensors with some
     non-zero amplitudes will be present.

   + 1D array of amplitudes for all time slices for the sensor with
     the specified ID.

   + 1D array of all sensor readings for a given time slice

   + A sequence of sensor IDs active in this peak. In the case of the
     PMTs, this will always be a sequence of consecutive integers
     starting at 0.

   + 1D array of ampltude sums over all time slices, with one entry
     for each sensor.

   + 1D array of amplitude sums over all sensors, with one entry for
     each time slice

**Comment by paolafer** [comment]
Great! I have a comment: I find it useful to be able to separate S1 and S2, because we use them in different ways, how is this done in  the new scheme?

**Comment by jacg** [comment]
@jahernando and @jjgomezcadenas requested something similar.

We are proposing 2 different, but isomorphic `Peak` types. The `peaks` property of `Signal`is to be upgraded to a method, which takes the type of peak you want as an optional argument:

    signal.peaks()       --> all peaks
    signal.peaks(S1) --> S1 peaks only
    signal.peaks(S2) --> S2 peaks only

**Comment by gonzaponte** [comment]
Actually, `Signal` already means "collection of peaks of a given type". This representation doesn't need to know which peaks you are referring to. The different interpretation comes from the user, not from its meaning. However, they can be combined in a (probably very useful) structure: a `PMap`. The structure of this class could be:
```
PMap

    s1s        : Signal
    s2s        : Signal
    evt_no     : int ?
    time_stamp : int ? 
    ...              ?
```

**Comment by paolafer** [comment]
Ok, so a PMap would contain a set of S1s and a set of S2s at the irene level, right? The selection of one and only one S1 stays after that.

**Comment by jjgomezcadenas** [comment]
Just change the name Signal by PMap, make Peak abstract (Peak_) and derive S1 and S2 from Peak_ to explicitly the types. That gives all the functionality we need

**Comment by jahernando** [comment]
S1 and S2 are structurally identical, but physically are very different. For me it is better to define them as types, S1, S2, inheriting from Peak_, and JJ proposes. If we rename Signal into PMaps with the functionality that Jack has added, then we have all the functionality, with a minimum, simple set of classes.

**Comment by gonzaponte** [comment]
Ok, we have peras-ed and:
1. We agree on `SensorResponses`.
3. We agree on the `Signal` to `PMap` renaming.
2. Do we want to have `S1` and `S2` to inherit from `Peak` or `Peak` to contain a `type` attribute (which would probably be implemented as a *Symbol*)?
4. Do we want `Signal` (now `PMap`) to have a `peaks(arg)` method that can provide all peaks or a subset of them depending on `arg` or do we prefer to have access to them as `s1_peaks`, `s2_peaks` (all of them being sequences of peaks)? I like the latter the most, but @jahernando pointed out that the former makes it easier for any other peak type to exist without breaking the interface, which is true.

Opinions, please.

**Comment by jjgomezcadenas** [comment]
>Do we want to have S1 and S2 to inherit from Peak or Peak to contain a type attribute (which would probably be implemented as a Symbol)?

Derive. They are different types.

> Do we want Signal (now PMap) to have a peaks(arg) method that can provide all peaks or a subset of them depending on arg or do we prefer to have access to them as s1_peaks, s2_peaks (all of them being sequences of peaks)? I like the latter the most, but @jahernando pointed out that the former makes it easier for any other peak type to exist without breaking the interface, which is true.

@jahernando is right, I think. I believe best is peaks (type), as suggested by Master @jacg 
peaks(S1), peaks(S2), peaks(AnyNewTypeDerivedFromPeaks) then works

**Comment by jacg** [comment]
Here is how we should proceed:

1. Create a pure python module (which should be called `pmaps` but that's problematic at the moment because there's already a `pmaps.pyx`. [Where is the record of the profiling procedure that indicatied that the classes in the latter needed Cythonizing?] So we should call it something like `pmaps_new.py` for now. This module should contain the classes set out above, with interfaces (and a bunch of `pass`es) only. Perhaps we can admit an outline implementation in the case of methods like `PMap.peaks()`, as some people struggle with understanding how this is supposed to work. This interface should be submitted in a PR which should be approved before we proceed. This PR is now open: #394.

2. The next step is to provide pure-Python implementations with tests. This must be approved in a second PR before we proceed any further.

3. Then we proceed to adapt all existing code that use PMaps, to the new interface.

4. Finally, we profile, and, **if and only if the profile suggests that it be necessary**, we Cythonize whatever needs to be Cythonized in `pmaps_new.py` which by now should be called simlpy `pmaps.py`.

Step 1. must be complete before @abotas and @gonzaponte arrive in Santiago, where they will start work on step 2.

Also please note that we want to avoid Cython modules with names that form part of the interface: the old `pmaps.pyx` should have been called `pmaps_c.pyx`(or maybe `_pmaps.pyx`) and its implementations imported into `pmaps.py` which is the name that clients want to use. The reason for this is that, with `pmaps.pyx`, anything that needs to be added to that module must be Cythonized immediately, or imported from a hidden pure-Python module: the former is evil, the latter is backwards.

In summary, this work will be carried out over 3 or 4 PRs. Before the 4th PR is started, I want to see evidence generated by profilers, proving that Cythonization is going to be worthwhile. Such evidence must be documented in an issue for all to see.

**Comment by gonzaponte** [comment]
This shouldn't be closed until we have the 4 PR approved, right?

**Comment by jacg** [comment]
This is about a new PMAPs *interface*. By approving #394 we have nailed down the new interface, so this can be (and therefore has been automaticallly) closed.

**Comment by abotas** [comment]
Do we want to add classes`SiPMResponses` and `PMTResponses` classes that inherit from `SensorResponses`? 

Currently we only have `SensorResponses` and its string representation does not know whether it contains the responses of PMTs or SiPMs. We could create `SiPMResponses` and `PMTResponses` that add self-documenting string representations.

**Comment by jjgomezcadenas** [comment]
Yes, I like the idea very much.

**Comment by abotas** [comment]
Ok sounds good
